File: java-command-api/src/main/java/bagay/commands/Commands.java
package bagay.commands;

import bagay.catalog.Catalog;
import bagay.chain.Chain;
import bagay.chain.Json;
import bagay.chain.Json.JsonValue;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * Java port of backend/commands.py's submit(): validate role rights and
 * lifecycle against the stream's own history, then produce the event to
 * append. Mirrors ARCHITECTURE.md invariant #3 -- this never reads a
 * projection to decide anything; the caller supplies the asset's current
 * status (obtained by replaying its stream, same as Python's
 * commands.load_asset) and the stream/chain heads (obtained from es_events /
 * es_chain_head / es_stream_heads, never from proj_*).
 *
 * What is intentionally NOT here yet: the actual JDBC transaction, the
 * SELECT ... FOR UPDATE-equivalent row lock, and idempotency-by-event_id.
 * Those are storage concerns for the PostgreSQL-wiring TODO item; this class
 * is the part that has to produce byte-identical hashes to Python regardless
 * of which storage layer eventually calls it.
 */
public final class Commands {

    private Commands() {
    }

    public static final DateTimeFormatter ISO_MILLIS =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC);

    /** Mirrors backend/commands.py ROLE_RIGHTS. SYSTEM may record anything. */
    public static final Map<String, Set<String>> ROLE_RIGHTS = Map.ofEntries(
            Map.entry("PUBLIC", Set.of("IssueReported")),
            Map.entry("FIELD_WORKER", Set.of("InspectionRecorded", "DamageRecorded",
                    "RepairStarted", "RepairCompleted")),
            Map.entry("SECRETARY", Set.of("AssetRegistered", "CustodyAssigned", "CustodyTransferred",
                    "IssueTriaged", "WorkOrderOpened", "EntryCorrected", "InspectionRecorded",
                    "DamageRecorded")),
            Map.entry("TREASURER", Set.of("RepairCompleted", "EntryCorrected")),
            Map.entry("KAGAWAD", Set.of("InspectionRecorded", "DamageRecorded", "HazardDeclared",
                    "WorkOrderOpened")),
            Map.entry("PUNONG_BARANGAY", Set.of("AssetDeclaredUnserviceable", "AssetDisposed",
                    "TurnoverSealed", "UserEnrolled", "UserDeactivated", "HazardDeclared",
                    "WorkOrderOpened")),
            Map.entry("VERIFIER", Set.of()),
            Map.entry("ADMIN", Set.of()),
            Map.entry("SYSTEM", Catalog.EVENTS.keySet())
    );

    public static final class Denied extends RuntimeException {
        public Denied(String message) {
            super(message);
        }
    }

    public static final class Conflict extends RuntimeException {
        public final String streamId;
        public final long currentVersion;

        public Conflict(String streamId, long currentVersion) {
            super(streamId + " is at version " + currentVersion);
            this.streamId = streamId;
            this.currentVersion = currentVersion;
        }
    }

    /** The two heads the command side is allowed to consult (never a projection). */
    public record ChainHead(long position, String headHash) {
    }

    public record StreamHead(long version, String headHash) {
    }

    /** What Commands.submit needs from the caller, and what it produces. */
    public record SubmitRequest(
            String streamId, String eventType, Json.JObject payload, String actorRole,
            String actorId, String source, Instant occurredAt, String currentAssetStatus,
            Long expectedVersion, String eventId) {
    }

    public record AppendedEvent(
            long globalPosition, String eventId, String streamId, long streamVersion,
            String eventType, int schemaVersion, String occurredAt, String recordedAt,
            String actorId, String actorRole, String trustLevel, String source,
            String prevHash, String streamPrevHash, String eventHash) {
    }

    public static String assetStream(String assetId) {
        return "asset:" + assetId;
    }

    /**
     * Validate, then compute the event to append. Pure: given the same heads
     * and the same asset status, always produces the same envelope and hash.
     * The caller is responsible for the actual INSERT + head updates, inside
     * one transaction, exactly as chain.append() does in Python.
     */
    public static AppendedEvent submit(SubmitRequest req, ChainHead chainHead, StreamHead streamHead) {
        if (!Catalog.EVENTS.containsKey(req.eventType())) {
            throw new Denied("unknown event type " + req.eventType());
        }
        Set<String> allowed = ROLE_RIGHTS.getOrDefault(req.actorRole(), Set.of());
        if (!allowed.contains(req.eventType())) {
            throw new Denied("role " + req.actorRole() + " may not record " + req.eventType());
        }

        boolean isAssetStream = req.streamId().startsWith("asset:");
        if (isAssetStream) {
            Catalog.Verdict verdict = Catalog.canApply(req.currentAssetStatus(), req.eventType());
            if (!verdict.ok()) {
                throw new Denied(verdict.reason());
            }
            if (req.expectedVersion() != null && req.expectedVersion() != streamHead.version()) {
                throw new Conflict(req.streamId(), streamHead.version());
            }
            // Same as Python: having checked it, the current stream version becomes
            // the expected version chain.append() will optimistically re-check.
        }

        Catalog.EventDef def = Catalog.EVENTS.get(req.eventType());
        String eventId = req.eventId() != null ? req.eventId() : UUID.randomUUID().toString();
        String occurredAt = ISO_MILLIS.format(truncateToMillis(req.occurredAt()));
        String recordedAt = ISO_MILLIS.format(truncateToMillis(Instant.now()));

        long globalPosition = chainHead.position() + 1;
        long streamVersion = streamHead.version() + 1;
        String prevHash = chainHead.headHash();
        String streamPrevHash = streamHead.headHash();

        Json.JObject envelope = buildEnvelope(eventId, req.streamId(), streamVersion, req.eventType(),
                1, occurredAt, recordedAt, req.actorId(), req.actorRole(), def.trust().name(),
                req.source(), null, null, req.payload(), Json.jarr(java.util.List.of()));

        String eventHash = Chain.computeHash(prevHash, streamPrevHash, globalPosition, envelope);

        return new AppendedEvent(globalPosition, eventId, req.streamId(), streamVersion, req.eventType(),
                1, occurredAt, recordedAt, req.actorId(), req.actorRole(), def.trust().name(),
                req.source(), prevHash, streamPrevHash, eventHash);
    }

    private static Instant truncateToMillis(Instant t) {
        return Instant.ofEpochMilli(t.toEpochMilli());
    }

    public static Json.JObject buildEnvelope(
            String eventId, String streamId, long streamVersion, String eventType, int schemaVersion,
            String occurredAt, String recordedAt, String actorId, String actorRole, String trustLevel,
            String source, String correlationId, String causationId, JsonValue payload,
            JsonValue attachments) {
        Json.JObject env = Json.jobj();
        env.members().put("event_id", Json.jstr(eventId));
        env.members().put("stream_id", Json.jstr(streamId));
        env.members().put("stream_version", Json.jnum(streamVersion));
        env.members().put("event_type", Json.jstr(eventType));
        env.members().put("schema_version", Json.jnum(schemaVersion));
        env.members().put("occurred_at", Json.jstr(occurredAt));
        env.members().put("recorded_at", Json.jstr(recordedAt));
        env.members().put("actor_id", Json.jstr(actorId));
        env.members().put("actor_role", Json.jstr(actorRole));
        env.members().put("trust_level", Json.jstr(trustLevel));
        env.members().put("source", Json.jstr(source));
        env.members().put("correlation_id", correlationId == null ? Json.jnull() : Json.jstr(correlationId));
        env.members().put("causation_id", causationId == null ? Json.jnull() : Json.jstr(causationId));
        env.members().put("payload", payload);
        env.members().put("attachments", attachments);
        return env;
    }
}

File: backend/commands.py
"""Command side: validate an intention against the asset's own history, then append.

The command side never reads projections to make a decision. It rebuilds the target
stream from the log, checks the lifecycle and the caller's role, and appends.
"""
from __future__ import annotations

import uuid

from . import catalog, chain, db

ROLE_RIGHTS = {
    "PUBLIC":           {"IssueReported"},
    "FIELD_WORKER":     {"InspectionRecorded", "DamageRecorded", "RepairStarted", "RepairCompleted"},
    "SECRETARY":        {"AssetRegistered", "CustodyAssigned", "CustodyTransferred", "IssueTriaged",
                         "WorkOrderOpened", "EntryCorrected", "InspectionRecorded", "DamageRecorded"},
    "TREASURER":        {"RepairCompleted", "EntryCorrected"},
    "KAGAWAD":          {"InspectionRecorded", "DamageRecorded", "HazardDeclared", "WorkOrderOpened"},
    "PUNONG_BARANGAY":  {"AssetDeclaredUnserviceable", "AssetDisposed", "TurnoverSealed",
                         "UserEnrolled", "UserDeactivated", "HazardDeclared", "WorkOrderOpened"},
    "VERIFIER":         set(),
    "ADMIN":            set(),
    "SYSTEM":           set(catalog.EVENTS),
}


class Denied(Exception):
    pass


def asset_stream(asset_id: str) -> str:
    return f"asset:{asset_id}"


def load_asset(conn: db.Conn, asset_id: str, as_of_position: int | None = None) -> dict:
    events = chain.stream_events(conn, asset_stream(asset_id), as_of_position)
    state = catalog.rebuild(events)
    state["asset_id"] = asset_id
    return state


def submit(conn: db.Conn, *, asset_id: str | None, event_type: str, payload: dict,
           actor_role: str, actor_id: str | None = None, expected_version: int | None = None,
           command_id: str | None = None, occurred_at=None, stream_id: str | None = None,
           source: str = "WEB") -> dict:
    """One entry point for every command. Returns the appended event row."""
    if event_type not in catalog.EVENTS:
        raise Denied(f"unknown event type {event_type}")
    if event_type not in ROLE_RIGHTS.get(actor_role, set()):
        raise Denied(f"role {actor_role} may not record {event_type}")

    stream = stream_id or asset_stream(asset_id)
    if stream.startswith("asset:"):
        state = load_asset(conn, asset_id)
        ok, why = catalog.can_apply(state, event_type)
        if not ok:
            raise Denied(why)
        if expected_version is not None and expected_version != state["version"]:
            raise chain.Conflict(stream, state["version"])
        expected_version = state["version"]

    trust, _ = catalog.EVENTS[event_type]
    return chain.append(
        conn, stream_id=stream, event_type=event_type, payload=payload,
        actor_id=actor_id or f"user:{actor_role.lower()}", actor_role=actor_role,
        trust_level=trust, source=source, occurred_at=occurred_at,
        expected_version=expected_version if stream.startswith("asset:") else None,
        event_id=command_id or str(uuid.uuid4()))


# --------------------------------------------------------------------------- helpers used by the UI and seed
def register_asset(conn, *, asset_id, name, category, purok, acquired_on=None,
                   acquisition_cost=0, fund_source=None, backfilled=False,
                   actor_role="SECRETARY", occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="AssetRegistered", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"asset_id": asset_id, "name": name, "category": category,
                           "location": {"purok": purok}, "acquired_on": acquired_on,
                           "acquisition_cost": acquisition_cost, "fund_source": fund_source,
                           "backfilled": backfilled})


def record_inspection(conn, *, asset_id, condition, findings="", actor_role="FIELD_WORKER",
                      occurred_at=None, command_id=None, expected_version=None):
    return submit(conn, asset_id=asset_id, event_type="InspectionRecorded", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id, expected_version=expected_version,
                  payload={"condition": condition, "findings": findings})


def record_damage(conn, *, asset_id, severity, service_disrupted=False, hazard_id=None,
                  estimated_cost=0, actor_role="FIELD_WORKER", occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="DamageRecorded", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"severity": severity, "service_disrupted": bool(service_disrupted),
                           "hazard_id": hazard_id, "estimated_cost": estimated_cost})


def report_issue(conn, *, asset_id, issue_category, description="", command_id=None, occurred_at=None):
    return submit(conn, asset_id=asset_id, event_type="IssueReported", actor_role="PUBLIC",
                  actor_id="public:qr", source="QR_PUBLIC", command_id=command_id,
                  occurred_at=occurred_at,
                  payload={"issue_id": str(uuid.uuid4()), "issue_category": issue_category,
                           "description": description})


def triage_issue(conn, *, asset_id, issue_id, decision, reason="", actor_role="SECRETARY",
                 occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="IssueTriaged", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"issue_id": issue_id, "decision": decision, "reason": reason})


def open_work_order(conn, *, asset_id, priority="P3", due_on=None, source_issue_id=None,
                    actor_role="SECRETARY", occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="WorkOrderOpened", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"work_order_id": str(uuid.uuid4()), "priority": priority,
                           "due_on": due_on, "source_issue_id": source_issue_id})


def start_repair(conn, *, asset_id, work_order_id, actor_role="FIELD_WORKER",
                 occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="RepairStarted", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"work_order_id": work_order_id})


def complete_repair(conn, *, asset_id, work_order_id, actual_cost=0, resulting_condition="GOOD",
                    fund_source="BARANGAY_DEVELOPMENT_FUND", actor_role="FIELD_WORKER",
                    occurred_at=None, command_id=None):
    return submit(conn, asset_id=asset_id, event_type="RepairCompleted", actor_role=actor_role,
                  occurred_at=occurred_at, command_id=command_id,
                  payload={"work_order_id": work_order_id, "actual_cost": actual_cost,
                           "resulting_condition": resulting_condition, "fund_source": fund_source})


def declare_hazard(conn, *, hazard_id, name, hazard_type, started_on, ended_on=None,
                   actor_role="PUNONG_BARANGAY", occurred_at=None, command_id=None):
    return submit(conn, asset_id=None, stream_id=f"hazard:{hazard_id}", event_type="HazardDeclared",
                  actor_role=actor_role, occurred_at=occurred_at, command_id=command_id,
                  payload={"hazard_id": f"hazard:{hazard_id}", "name": name,
                           "hazard_type": hazard_type, "started_on": started_on, "ended_on": ended_on})


def seal_turnover(conn, *, psgc, outgoing_term, incoming_term, witnesses, actor_role="PUNONG_BARANGAY"):
    cp = chain.create_checkpoint(conn, reason="TURNOVER")
    assets = conn.one("SELECT COUNT(*) AS c FROM proj_asset_current")
    return submit(conn, asset_id=None, stream_id=f"barangay:{psgc}", event_type="TurnoverSealed",
                  actor_role=actor_role,
                  payload={"checkpoint_position": cp["position"], "checkpoint_hash": cp["head_hash"],
                           "outgoing_term": outgoing_term, "incoming_term": incoming_term,
                           "asset_count": (assets or {}).get("c", 0), "witnesses": witnesses})
